pkgs/os-specific/linux/zfs/default.nix

```diff
@@ -202,7 +202,21 @@ let
         homepage = "https://github.com/openzfs/zfs";
         changelog = "https://github.com/openzfs/zfs/releases/tag/zfs-${version}";
         license = lib.licenses.cddl;
-        platforms = lib.platforms.linux;
+
+        # As of zfs-2.1.5, the case which sets TARGET_CPU has branches for only
+        # five CPU families:
+        #
+        #   https://github.com/openzfs/zfs/blob/6a6bd493988c75331deab06e5352a9bed035a87d/config/always-arch.m4#L16
+        #
+        # It is important that `meta.availableOn` fails for platforms where zfs
+        # does not build, because this is the only way to produce a NixOS
+        # installer on those platforms (the `boot.zfs.enabled` property is
+        # `readOnly`).
+        #
+        platforms =
+          with lib.systems.inspect.patterns;
+          map (p: p // isLinux) [ isx86_32 isx86_64 isPower isAarch64 isSparc ];
```

Comment:
This is a bit complicated.  There aren't that many supported doubles, so couldn't we just list them?

Reply:
IMHO platform doubles are sort of a crude last resort, like colon-separated $PATH strings instead of a list of strings, but maybe that's just my personal taste.

Once you expand out all the 32-bit/64-bit and endianness possibilities the list gets longer since the ZFS configury uses shell globbing for this.  Latest push c5138fd1faeb does this.  If you like it it's okay with me; if not let me know and I will roll it back.

Reply:
Hmm.  Did the previous version definitely work?  I don't see how it would since the check is implemented by just checking if a system double is in the platforms list.

https://github.com/NixOS/nixpkgs/blob/987d2f575aff1da36d964a6e3fd3e4ba42445a76/pkgs/stdenv/generic/check-meta.nix#L57-L59

If it _did_ work, I guess it's fine, although I'd like to understand how.  Otherwise, you could do something like this:

```nix
with platforms;
  intersectLists linux (x86 ++ power ++ arm ++ aarch64 ++ sparc);
```

Although I notice we're missing `lib.platforms.power` and `lib.platforms.sparc`, so you'd have to add those in lib/systems/doubles.nix.

Reply:
> Did the previous version definitely work?

Yes.

I assume "previous version" means 69c266e4fb730dee38923fcea7012c0743f2a2ec.

> I don't see how it would since the check is implemented by just checking if a system double is in the platforms list.

> If it _did_ work, I guess it's fine, although I'd like to understand how.

Since #37395 you can now pass an attrset to `meta.platformMatch` and it does a structural match on it.  `meta.availableOn` uses that codepath.

It looks like `check-meta.nix` was not updated, so I have opened https://github.com/NixOS/nixpkgs/pull/195294

OfBorg CI passed because the `badPlatforms` check is looking for *mis*matches, so it "fails open" on unexpected input like wrong-type or misspelled platform strings (another reason to move away from them...)

Reply:
Oh cool, I've never seen this before.